worker/config/config.go

```diff
@@ -127,6 +134,20 @@ func (c *Config) Defaults() error {
 		return fmt.Errorf("must specify local registry directory")
 	}
 
+	// mem usage limits for creating sandboxes
+	if c.Create_sandbox_soft_limit == 0.0 {
+		c.Create_sandbox_soft_limit = 0.5
+	} 
+
+	if c.Create_sandbox_hard_limit == 0.0 {
+		c.Create_sandbox_hard_limit = 0.9
+	} 
+
+	// max number of sandboxes being created at the same time (after hit soft limit)
+	if c.Max_num_sandbox_creating == 0 {
+		c.Max_num_sandbox_creating = 4
```

Comment:
What's the rational for this limit?

Reply:
We are thinking to reduce the concurrency of spinning up new sandboxes once the memory usage hits `Create_sandbox_soft_limit`.

Suppose without `Create_sandbox_soft_limit ` and `Max_num_sandbox_creating `, when the memory usage is close to the `Create_sandbox_hard_limit `, a bunch of incoming requests would detect that the memory usage is lower than the hard limit, thus try to spin up multiple sandboxes concurrently, leading to a surge in memory usage and crash the system.

In addition, this limit is configurable in twofold: It is configurable in itself, plus that once the soft limit is set to be greater than or equal to the hard limit, it does not take effect anymore.

Reply:
Let me paraphrase what I think you're saying: mem measurements are only updated every second, so we don't want to risk letting too many in at once, shooting past the limit before we re-measure.

If that's correct, I think the correct thing to do is limit the rate, not the concurrency.

How about we do this:
1. estimate the memory consumption of the new sandbox (based on a config, or average of all other lambdas)
2. update the mem-usage snapshot based on our estimate (it will be exact in the next second)
3. wait if starting would bump our mem estimate past the threshold

Reply:
Your understanding is correct.

Your proposal makes a lot of sense, though we think it may cause in some cases, for example: suppose a sandbox creating request comes, with the mem usage is right under hard limit, we let it spin up a sandbox and "update the mem-usage snapshot based on our estimate" over the hard limit, then after one sec another sandbox creating request comes, it will remeasure the exact mem usage -- if the first sandbox is still being created and takes little memory, then this request will be approved as well, leading mem usage to past the hard limit.

To resolve this, we are thinking to keep the [`hms.numSbCreating`](https://github.com/open-lambda/open-lambda/pull/54/files#diff-3d8711386f9437d690ff9ee2452d22b2R353) counter to keep track of how many sandboxes are being created, and approve/queue a request based on [whether the sum of host mem usage and estimated mem usage of all sandboxes being created is over the hard limit](https://github.com/open-lambda/open-lambda/pull/54/files#diff-3d8711386f9437d690ff9ee2452d22b2R344).

We have updated this in https://github.com/open-lambda/open-lambda/pull/54.